File: imageblender.py
#!/usr/bin/python

import sys
from itertools import *

def grouper(iterable, n, fillvalue=None):
    args = [iter(iterable)] * n
    return izip_longest(*args, fillvalue=fillvalue)

def rgb_to_bin(r,g,b):
    return int(hex(r)[2:] + hex(g)[2:] + hex(b)[2:], 16)

def bin_to_rgb(b):
    h = '%06X'%(b)
    return map(lambda x:int(''.join(x), 16), grouper(h, 2))

def pnm_to_bin(p):
    w,h = map(int,p[1].split(' '))
    data = map(int, ' '.join(p[3:]).replace('\n', ' ').split())
    bin = []
    lines = grouper(data, w*3)
    for line in lines:
        data = []
        for rgb in grouper(line, 3):
            data.append(rgb_to_bin(*rgb))
        bin.append(data)
    return bin

def bin_to_pnm(b):
    pnm = 'P3 {} {} 255 '.format(len(b[0]), len(b))
    b = [z for a in b for z in a]
    pnm += ' '.join(map(lambda x:' '.join(map(str, bin_to_rgb(x))), b))
    return pnm

def imageblender(img):
    h = len(img)
    w = len(img[0])
    for y in range(w):
        for x in range(h):
            val = (img[x][y] + img[x][~y] + img[~x][y] + img[~x][~y])//4
            img[x][y],img[x][~y],img[~x][y],img[~x][~y] = (val,)*4
    return img

def main(fname):
    bin = pnm_to_bin(open(fname).read().split('\n'))
    bin = imageblender(bin)
    return bin_to_pnm(bin)

if __name__ == '__main__':
    print main(sys.argv[1])

File: js/ImageBlender.js
var grouper = function (iterable, n, fillvalue) {
	var extra = iterable.length % n,
		result = [];
	while (extra) {
		iterable.push(fillvalue);
		extra--;
	}
	while (iterable.length) {
		elem = [];
		while (elem.length < n) {
			elem.push(iterable.pop())
		}
		elem.reverse()
		result.push(elem);
	}
	result.reverse()
	return result;
}

var flatten = function (lst) {
	var result = [];
	for (var i in lst) {
		if (!(lst[i] instanceof Array)) {
			result.push([lst[i]]);
		}
		else {
			result.push(flatten(lst[i]));
		}
	}
	return result.reduce(function (i, j) {
		return i.concat(j);
	});
}

var pnm_to_bin = function (p) {
	var w = parseInt(p[1].split(" ")[0]),
		h = parseInt(p[1].split(" ")[1]),
		data = p.slice(3).join(" ").replace("\n", " ").split(/\s+/g),
		bin = [];
	data = data.filter(function (i) {
		return i.length;
	}).map(function (i) {
		return parseInt(i);
	});
	var lines = grouper(data, w*3),
		i = 0;
	while (i < lines.length) {
		var lgroup = grouper(lines[i], 3),
			d = [],
			j = 0;
		while (j < lgroup.length) {
			d.push(lgroup[j]);
			j++;
		}
		bin.push(d);
		i++;
	}
	return bin;
}

var bin_to_pnm = function (b) {
	pnm = "P3 " + b[0].length + " " + b.length + " 255 ";
	b = flatten(b);
	pnm += b.map(function (i) {
		return i.toString();
	}).join(" ");
	return pnm;
}

var imageblender = function (img) {
	var nimg = new Array() .concat(img);
	var h = nimg.length,
		w = nimg[0].length,
		y = 0;
	while (y < w) {
		var x = 0;
		while (x < h) {
			var i = 0;
			while (i < 3) {
				var val = Math.floor((img[x][y][i] + img[x][w+~y][i] + img[h+~x][y][i] + img[h+~x][w+~y][i]) / 4);
				nimg[x][y][i] = val;
				nimg[x][w+~y][i] = val;
				nimg[h+~x][y][i] = val;
				nimg[h+~x][w+~y][i] = val;
				i++;
			}
			x++;
		}
		y++;
	}
	return nimg;
}

var main = function (content) {
	b = pnm_to_bin(content.split("\n"));
	bin = imageblender(b);
	return bin_to_pnm(bin);
}